packages/headless/src/features/insight-user-actions/insight-user-actions-preprocessing.ts

```diff
@@ -0,0 +1,260 @@
+import {
+  UserAction,
+  UserActionsState,
+  UserActionTimeline,
+  UserActionType,
+  UserSession,
+} from './insight-user-actions-state';
+
+const MSEC_IN_SECOND = 1000;
+const SECONDS_IN_MINUTE = 60;
+const MAX_MINUTES_IN_SESSION = 30;
+const MAX_MSEC_IN_SESSION =
+  MAX_MINUTES_IN_SESSION * SECONDS_IN_MINUTE * MSEC_IN_SECOND;
+
+interface rawUserAction {
+  name: string;
+  value: string;
+  time: string;
+}
+
+export const preprocessActionsData = (
```

Comment:
Split this function into two distinct functions (and nothing else):
 - One that computes the proper actions: filtered, sorted and mapped (and in a single iteration)
 - One that computes timelines from the proper actions (this should be doable in a single iteration on the filtered action as well, as the discriminating factor is the time delta between two actions)

Reply:
We can't filter before building the sessions because of the case where a filtered action would be within a session and if filtered would cause 1 session to be broken into 2 sessions.

For example:

action 1 is at 2h00 pm

action 2 is at 2h29 pm (action to be excluded)

action 3 is at 2h31 pm 

If we filter first, it would result in 2 sessions where in reality action 1 and 3 are really in the same session.

Reply:
@louis-bompart I can't figure out how we could do it in only two iteration though.

1. We need to sort by inverse chronological order (I don't think we can trust that the API will return the actions in chronological order or even inverted, so we need to sort ourselves.)
2. We need to split the actions into sessions (this cannot be done before the filtering, because the "filtered" actions could be the actions that tie a session together, a session is a group of actions where there's no more than 30 minutes of inactivity).
3. We need to map and filter to remove the custom events and empty search events.
4. We need to organize all these sessions afterwards around a specific datetime.

So minimum I can figure out is 
1 to sort, I think JS sort is `nlog(n)` IIRC.
1 iteration to split into sessions, probably can map at the same time (but I think we might not want to, explained next) O(n)
1 operation to find the session belonging to the case creation datetime (worst case O(n)? if we can't find any)
After that, we're now dealing with a much smaller dataset since we're keeping at maximum 5 sessions worth of actions,
This is when we could map and filter which would be O(n) but a much lower `n` than the original number of actions?

What do you think? is it better to map at the same time as sort or split into sessions for example, which would mean we would map a lot of actions we don't actually care about? Or to map at the end when we also are filtering but working on a much smaller collection of data?

Reply:
👍 had some wrong assumptions, but yeah you captured the essence of my comment: too many iterations, tone it down.